step/step.go

```diff
@@ -0,0 +1,68 @@
+package step
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/bitrise-io/go-steputils/v2/cache"
+	"github.com/bitrise-io/go-steputils/v2/stepconf"
+	"github.com/bitrise-io/go-utils/v2/env"
+	"github.com/bitrise-io/go-utils/v2/log"
+)
+
+const stepId = "restore-gradle-cache"
+
+// Cache key templates
+// OS + Arch: to guarantee that stack-specific content (absolute paths, binaries) are stored separately
+// checksum values:
+// - `**/*.gradle*`: Gradle build files in any submodule, including ones written in Kotlin (*.gradle.kts)
+// - `**/gradle-wrapper.properties`: contains exact Gradle version
+// - `gradle.properties`: contains Gradle config values
+// - `gradle/libs.versions.toml`: version catalog file, contains dependencies and their versions
+var keys = []string{
+	`{{ .OS }}-{{ .Arch }}-gradle-cache-{{ checksum "**/*.gradle*" "**/gradle-wrapper.properties" "gradle.properties" "gradle/libs.versions.toml" }}`,
+	`{{ .OS }}-{{ .Arch }}-gradle-cache-`,
+}
+
+type Input struct {
+	Verbose bool `env:"verbose,required"`
+}
+
+type RestoreCacheStep struct {
+	logger      log.Logger
+	inputParser stepconf.InputParser
+	envRepo     env.Repository
+}
+
+func New(
+	logger log.Logger,
+	inputParser stepconf.InputParser,
+	envRepo env.Repository,
+) RestoreCacheStep {
+	return RestoreCacheStep{
```

Comment:
I'm always curious about memory allocations.  This likely is pretty small and the compiler might actually identify this as an "escapable" or inlineable function, so it might be auto-optimized or a complete non-issue.  In the worst case scenario, this would allocate a `RestoreCacheStep` and then immediate copy it into the stack provided by the caller, right?  That only happens once per step and is unlikely to be any kind of performance concern, but it triggered my curiosity around what's happening under the hood. :)

Reply:
Good question, and I couldn't find a definitive answer, but I wouldn't worry it in this case, as you say it only happens once per step :)